关于parallelism config 和 quantize config共同使用的问题

同时使用并行和量化一般是显存有限的情况下，但是我看实现过程是先做的并行再做的量化，但如果我显存能放得下我肯定不需要做量化了。就是因为放不下所以才希望通过量化降低显存，再tp拆分到两张卡上并行计算。

但是试了一下先量化再并行，发现pytorch原生的TP不支持量化后TP。所以我尝试了下下面这个代码：
```python
import torch.distributed as dist
def init_torch_distributed():
    rank = int(os.environ.get("RANK", 0))
    local_rank = int(os.environ.get("LOCAL_RANK", 0))
    world_size = int(os.environ.get("WORLD_SIZE", 1))
    master_addr = os.environ.get("MASTER_ADDR", "127.0.0.1")
    master_port = os.environ.get("MASTER_PORT", "29500")

    if torch.cuda.is_available():
        torch.cuda.set_device(local_rank)

    if world_size > 1:
        try:
            dist.init_process_group(
                backend="nccl",
                init_method=f"tcp://{master_addr}:{master_port}",
                rank=rank,
                world_size=world_size,
            )
            print(f"[INFO] DDP init success: rank={rank}/{world_size}, local_rank={local_rank}")
        except Exception as e:
            print(f"[WARN] DDP init failed: {e}, fallback to single-GPU")
            os.environ["RANK"] = "0"
            os.environ["LOCAL_RANK"] = "0"
            os.environ["WORLD_SIZE"] = "1"
            world_size = 1
    else:
        print("[INFO] Single-GPU mode, skip DDP init")
    return world_size

from diffusers import QwenImageEditPlusPipeline
from os.path import join as opj
import cache_dit
from cache_dit import ParallelismConfig,QuantizeConfig

init_torch_distributed()

model_path = "./Qwen-Image-Edit-2509")

parallelism_config = ParallelismConfig(tp_size=2,extra_parallel_modules=["text_encoder"])
quantize_config = QuantizeConfig(quant_type="float8")

pipe = QwenImageEditPlusPipeline.from_pretrained(
            model_path,
            torch_dtype=torch.bfloat16,
        )
from cache_dit.parallelism.dispatch import enable_parallelism

pipe.transformer = enable_parallelism(pipe.transformer,parallelism_config=parallelism_config)
cache_dit.enable_cache(pipe,quantize_config = quantize_config)
from cache_dit.parallelism.text_encoders import maybe_enable_parallelism_for_text_encoder
pipe.text_encoder = maybe_enable_parallelism_for_text_encoder(pipe.text_encoder,parallelism_config=parallelism_config)
pipe.to(torch.cuda.current_device())
pipe.vae.enable_tiling()

image = Image.fromarray(load_image("https://huggingface.co/datasets/huggingface/documentation-images/resolve/main/diffusers/yarn-art-pikachu.png"))
prompt = ("Make Pikachu hold a sign that says 'Qwen Edit is awesome', yarn art style, detailed, vibrant colors")
image = pipe(image, prompt, num_inference_steps=50).images[0]

```
但这个代码会报错：
```python
[rank0]: Traceback (most recent call last):
[rank0]:   File "/media/74nvme/software/miniconda3/envs/quant_compile/lib/python3.11/site-packages/torch/distributed/tensor/_dispatch.py", line 156, in dispatch
[rank0]:     self.sharding_propagator.propagate(op_info)
[rank0]:   File "/media/74nvme/software/miniconda3/envs/quant_compile/lib/python3.11/site-packages/torch/distributed/tensor/_sharding_prop.py", line 327, in propagate
[rank0]:     OutputSharding, self.propagate_op_sharding(op_info.schema)
[rank0]:                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
[rank0]:   File "/media/74nvme/software/miniconda3/envs/quant_compile/lib/python3.11/site-packages/torch/distributed/tensor/_sharding_prop.py", line 46, in __call__
[rank0]:     return self.cache(*args, **kwargs)
[rank0]:            ^^^^^^^^^^^^^^^^^^^^^^^^^^^
[rank0]:   File "/media/74nvme/software/miniconda3/envs/quant_compile/lib/python3.11/site-packages/torch/distributed/tensor/_sharding_prop.py", line 346, in propagate_op_sharding_non_cached
[rank0]:     op_strategy = self.op_strategy_funcs[op_schema.op](strategy_schema)
[rank0]:                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
[rank0]:   File "/media/74nvme/software/miniconda3/envs/quant_compile/lib/python3.11/site-packages/torch/distributed/tensor/_ops/_matrix_ops.py", line 248, in scaled_mm_strategy
[rank0]:     return _scaled_mm_like_strategy("mk,kn->mn", mesh, op_schema)
[rank0]:            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
[rank0]:   File "/media/74nvme/software/miniconda3/envs/quant_compile/lib/python3.11/site-packages/torch/distributed/tensor/_ops/_matrix_ops.py", line 168, in _scaled_mm_like_strategy
[rank0]:     assert bias_strategy is None, "_scaled_mm on DTensors doesn't support bias"
[rank0]:            ^^^^^^^^^^^^^^^^^^^^^
[rank0]: AssertionError: _scaled_mm on DTensors doesn't support bias

[rank0]: The above exception was the direct cause of the following exception:

[rank0]: Traceback (most recent call last):
[rank0]:   File "/media/242hdd/test.py", line 242, in <module>
[rank0]:     image = pipe(image, prompt, num_inference_steps=8).images[0]
[rank0]:             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
[rank0]:   File "/media/74nvme/software/miniconda3/envs/quant_compile/lib/python3.11/site-packages/torch/utils/_contextlib.py", line 120, in decorate_context
[rank0]:     return func(*args, **kwargs)
[rank0]:            ^^^^^^^^^^^^^^^^^^^^^
[rank0]:   File "/media/74nvme/software/miniconda3/envs/quant_compile/lib/python3.11/site-packages/diffusers/pipelines/qwenimage/pipeline_qwenimage_edit_plus.py", line 806, in __call__
[rank0]:     noise_pred = self.transformer(
[rank0]:                  ^^^^^^^^^^^^^^^^^
[rank0]:   File "/media/74nvme/software/miniconda3/envs/quant_compile/lib/python3.11/site-packages/torch/nn/modules/module.py", line 1775, in _wrapped_call_impl
[rank0]:     return self._call_impl(*args, **kwargs)
[rank0]:            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
[rank0]:   File "/media/74nvme/software/miniconda3/envs/quant_compile/lib/python3.11/site-packages/torch/nn/modules/module.py", line 1786, in _call_impl
[rank0]:     return forward_call(*args, **kwargs)
[rank0]:            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
[rank0]:   File "/media/74nvme/software/miniconda3/envs/quant_compile/lib/python3.11/site-packages/diffusers/utils/peft_utils.py", line 315, in wrapper
[rank0]:     result = forward_fn(self, *args, **kwargs)
[rank0]:              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
[rank0]:   File "/media/74nvme/software/miniconda3/envs/quant_compile/lib/python3.11/site-packages/diffusers/models/transformers/transformer_qwenimage.py", line 953, in forward
[rank0]:     encoder_hidden_states, hidden_states = block(
[rank0]:                                            ^^^^^^
[rank0]:   File "/media/74nvme/software/miniconda3/envs/quant_compile/lib/python3.11/site-packages/torch/nn/modules/module.py", line 1775, in _wrapped_call_impl
[rank0]:     return self._call_impl(*args, **kwargs)
[rank0]:            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
[rank0]:   File "/media/74nvme/software/miniconda3/envs/quant_compile/lib/python3.11/site-packages/torch/nn/modules/module.py", line 1786, in _call_impl
[rank0]:     return forward_call(*args, **kwargs)
[rank0]:            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
[rank0]:   File "/media/74nvme/software/miniconda3/envs/quant_compile/lib/python3.11/site-packages/diffusers/models/transformers/transformer_qwenimage.py", line 675, in forward
[rank0]:     img_mod_params = self.img_mod(temb)  # [B, 6*dim]
[rank0]:                      ^^^^^^^^^^^^^^^^^^
[rank0]:   File "/media/74nvme/software/miniconda3/envs/quant_compile/lib/python3.11/site-packages/torch/nn/modules/module.py", line 1775, in _wrapped_call_impl
[rank0]:     return self._call_impl(*args, **kwargs)
[rank0]:            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
[rank0]:   File "/media/74nvme/software/miniconda3/envs/quant_compile/lib/python3.11/site-packages/torch/nn/modules/module.py", line 1786, in _call_impl
[rank0]:     return forward_call(*args, **kwargs)
[rank0]:            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
[rank0]:   File "/media/74nvme/software/miniconda3/envs/quant_compile/lib/python3.11/site-packages/torch/nn/modules/container.py", line 250, in forward
[rank0]:     input = module(input)
[rank0]:             ^^^^^^^^^^^^^
[rank0]:   File "/media/74nvme/software/miniconda3/envs/quant_compile/lib/python3.11/site-packages/torch/nn/modules/module.py", line 1775, in _wrapped_call_impl
[rank0]:     return self._call_impl(*args, **kwargs)
[rank0]:            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
[rank0]:   File "/media/74nvme/software/miniconda3/envs/quant_compile/lib/python3.11/site-packages/torch/nn/modules/module.py", line 1881, in _call_impl
[rank0]:     return inner()
[rank0]:            ^^^^^^^
[rank0]:   File "/media/74nvme/software/miniconda3/envs/quant_compile/lib/python3.11/site-packages/torch/nn/modules/module.py", line 1829, in inner
[rank0]:     result = forward_call(*args, **kwargs)
[rank0]:              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
[rank0]:   File "/media/74nvme/software/miniconda3/envs/quant_compile/lib/python3.11/site-packages/torch/nn/modules/linear.py", line 134, in forward
[rank0]:     return F.linear(input, self.weight, self.bias)
[rank0]:            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
[rank0]:   File "/media/74nvme/software/miniconda3/envs/quant_compile/lib/python3.11/site-packages/torchao/utils.py", line 655, in _dispatch__torch_function__
[rank0]:     return cls._TORCH_FN_TABLE[cls][func](func, types, args, kwargs)
[rank0]:            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
[rank0]:   File "/media/74nvme/software/miniconda3/envs/quant_compile/lib/python3.11/site-packages/torchao/utils.py", line 458, in wrapper
[rank0]:     return _func(f, types, args, kwargs)
[rank0]:            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
[rank0]:   File "/media/74nvme/software/miniconda3/envs/quant_compile/lib/python3.11/site-packages/torchao/quantization/quantize_/workflows/float8/float8_tensor.py", line 265, in _
[rank0]:     return _float8_addmm_impl(input_tensor, weight_tensor.t(), bias)
[rank0]:            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
[rank0]:   File "/media/74nvme/software/miniconda3/envs/quant_compile/lib/python3.11/site-packages/torchao/quantization/quantize_/workflows/float8/float8_tensor.py", line 432, in _float8_addmm_impl
[rank0]:     res = addmm_float8_unwrapped_inference(
[rank0]:           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
[rank0]:   File "/media/74nvme/software/miniconda3/envs/quant_compile/lib/python3.11/site-packages/torchao/float8/inference.py", line 113, in addmm_float8_unwrapped_inference
[rank0]:     return torch._scaled_mm(
[rank0]:            ^^^^^^^^^^^^^^^^^
[rank0]:   File "/media/74nvme/software/miniconda3/envs/quant_compile/lib/python3.11/site-packages/torch/_compile.py", line 53, in inner
[rank0]:     return disable_fn(*args, **kwargs)
[rank0]:            ^^^^^^^^^^^^^^^^^^^^^^^^^^^
[rank0]:   File "/media/74nvme/software/miniconda3/envs/quant_compile/lib/python3.11/site-packages/torch/_dynamo/eval_frame.py", line 1044, in _fn
[rank0]:     return fn(*args, **kwargs)
[rank0]:            ^^^^^^^^^^^^^^^^^^^
[rank0]:   File "/media/74nvme/software/miniconda3/envs/quant_compile/lib/python3.11/site-packages/torch/distributed/tensor/_api.py", line 349, in __torch_dispatch__
[rank0]:     return DTensor._op_dispatcher.dispatch(
[rank0]:            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
[rank0]:   File "/media/74nvme/software/miniconda3/envs/quant_compile/lib/python3.11/site-packages/torch/distributed/tensor/_dispatch.py", line 169, in dispatch
[rank0]:     raise RuntimeError(
[rank0]: RuntimeError: Sharding propagation failed for Op(op=aten._scaled_mm.default, args_schema=Spec(R on (1, 3072)), Spec(S(1) on (3072, 18432)), Spec(R on (1, 1)), Spec(R on (1, 1)), Spec(S(0) on (18432,)), None, torch.bfloat16, True @ mesh: (2,))
[rank0]:[W312 15:56:37.734328590 ProcessGroupNCCL.cpp:1524] Warning: WARNING: destroy_process_group() was not called before program exit, which can leak resources. For more info, please see https://pytorch.org/docs/stable/distributed.html#shutdown (function operator())
```
我是在两张4090上测试的

--- a/src/cache_dit/quantize/torchao/quantize_ao.py
+++ b/src/cache_dit/quantize/torchao/quantize_ao.py
@@ -119,11 +119,23 @@ def quantize_ao(
                 return False
 
             # check blockwise fp8 support for linear layers, if not supported, skip quantization for that layer
-            if quant_type == "fp8_blockwise" and not _check_blockwise_fp8_support(m):
+            if quant_type == "fp8_blockwise" and not _check_if_linear_fp8_blockwise_can_support(m):
                 weight_shape = tuple(m.weight.shape)
                 if verbose:
                     logger.info(
                         f"Skip Quantization: {name} -> pattern<w{weight_shape} % block_size(128, 128) != 0>"
+                    )
+                num_skip_linear += 1
+                return False
+
+            if quant_type in [
+                "fp8_w8a8_dq",
+                "fp8_blockwise",
+            ] and not _check_if_linear_with_bias_fp8_can_support(m):
+                if verbose:
+                    logger.info(
+                        f"Skip Quantization: {name} -> "
+                        f"pattern<DTensor + bias is not supported for _scaled_mm>"
                     )
                 num_skip_linear += 1
                 return False
@@ -285,7 +297,7 @@ def quantize_ao(
     return module
 
 
-def _check_blockwise_fp8_support(module: torch.nn.Linear):
+def _check_if_linear_fp8_blockwise_can_support(module: torch.nn.Linear) -> bool:
     try:
         from torchao.quantization.utils import get_block_size
         from torchao.quantization import PerBlock
@@ -313,3 +325,25 @@ def _check_blockwise_fp8_support(module: torch.nn.Linear):
     except Exception as e:
         logger.debug(f"Failed to get block size for module {module}: {e}")
         return False
+
+
+def _check_if_linear_with_bias_fp8_can_support(module: torch.nn.Linear) -> bool:
+    # Avoid: AssertionError("_scaled_mm on DTensors doesn't support bias")
+    # Check we are in distributed environment and the linear layer has bias,
+    # and if the weight is DTensor, if all conditions are met, we will skip
+    # quantization for that layer to avoid potential issues.
+    if not torch.distributed.is_initialized():
+        return True
+    # If the linear layer doesn't have bias, we can quantize it without issues.
+    if not hasattr(module, "bias") or module.bias is None:
+        return True
+    # For the case where the linear layer has bias, we need to check if the weight
+    # or bias is DTensor. We only quantize the linear layer when both weight and
+    # bias are not DTensor.
+    from torch.distributed._tensor import DTensor
+
+    weight_tensor = getattr(module, "weight", None)
+    bias_tensor = getattr(module, "bias", None)
+    if weight_tensor is None or bias_tensor is None:
+        return False
+    return not isinstance(weight_tensor, DTensor) and not isinstance(bias_tensor, DTensor)

